src/number_theory/fermat_psp.lean

```diff
@@ -0,0 +1,578 @@
+    exact_mod_cast h₂ }
+end
+
+/--
+If `n` is a Fermat pseudoprime to base `b`, then `n` is coprime with `b`, assuming that `b` is
+positive.
+
+This lemma is a small wrapper based on `coprime_of_probable_prime`
+-/
+lemma coprime_of_fermat_psp (n b : ℕ) (h : fermat_psp n b) (h₁ : 1 ≤ b) : nat.coprime n b :=
+begin
+  cases h with hp hn,
+  cases hn with hn₁ hn₂,
+  exact coprime_of_probable_prime n b hp (by linarith) h₁,
+end
+
+/--
+All composite numbers are Fermat pseudoprimes to base 1.
+-/
+lemma pseudoprime_of_base_one (n : ℕ) (h₁ : 1 < n) (h₂ : ¬nat.prime n) : fermat_psp n 1 :=
+begin
+  split,
+  { have h : 0 = 1^(n - 1) - 1 := by norm_num,
+    show n ∣ 1^(n - 1) - 1, from h ▸ (dvd_zero n) },
+  { exact ⟨h₂, h₁⟩ }
+end
+
+-- For lemmas that are needed to prove statements in this file, but aren't directly related to
+-- Fermat pseudoprimes
+section helper_lemmas
+
+private lemma odd_of_prime_gt_two (p : ℕ) (h : nat.prime p) (hp : 2 < p) : odd p :=
+begin
+  rw [nat.odd_iff_not_even, even_iff_two_dvd],
+  rw nat.prime_dvd_prime_iff_eq nat.prime_two h,
+  exact ne_of_lt hp
+end
+
+private lemma odd_pow_lem (a : ℤ) (n k : ℕ) (h₁ : k ∣ n) (h₂ : odd (n / k)) : a^k + 1 ∣ a^n + 1 :=
+begin
+  -- Let m be the natural number such that k * m = n. Then (-1)^m = -1 since m is odd by hypothesis.
+  generalize h₃ : n / k = m,
+  have hm : k * m = n := by { rw [←h₃, mul_comm], exact nat.div_mul_cancel h₁ },
+  have hm₁ : (-1 : ℤ)^m = -1 := odd.neg_one_pow (h₃ ▸ h₂),
+
+  -- a^k ≡ -1 (mod a^k + 1)
+  have hk : a^k + 1 ∣ a^k + 1 - 1 -(-1) := by norm_num,
+  have hk₁ : a^k + 1 - 1 ≡ -1 [ZMOD a^k + 1] := (int.modeq_of_dvd hk).symm,
+  have hk₂ : a^k ≡ -1 [ZMOD a^k + 1] := by rwa int.add_sub_cancel at hk₁,
+
+  -- a^n = (a^k)^m ≡ (-1)^m = -1 (mod a^k + 1)
+  have ha : a^n = (a^k)^m := by rw [←pow_mul, hm],
+  have ha₂ : (a^k)^m ≡ (-1)^m [ZMOD a^k + 1] := int.modeq.pow m hk₂,
+  have ha₃ : a^n ≡ (-1)^m [ZMOD a^k + 1] := (eq.symm ha) ▸ ha₂,
+  have ha₄ : a^n ≡ (-1) [ZMOD a^k + 1] := hm₁ ▸ ha₃,
+
+  -- Therefore, a^k + 1 divides a^n + 1
+  show a^k + 1 ∣ a^n + 1, from (int.modeq.symm ha₄).dvd
+end
+
+private lemma sub_one_dvd_pow_sub_one (a n : ℕ) (h : a ≥ 1) : a - 1 ∣ a^n - 1 :=
+begin
+  induction n with n ih,
+  { simp },
+  { have h₁ : a - 1 ∣ a * (a^n - 1) := dvd_mul_of_dvd_right ih _,
+    have h₂ : 1 ≤ a ^ n := nat.one_le_pow n a (show 0 < a, by linarith),
+    have h₃ : 1 ≤ a^(n + 1) := by { rw pow_succ, nlinarith },
+    have h₄ : a * (a ^ n - 1) + (a - 1) = a ^ (n + 1) - 1, by { zify, ring_exp },
+    exact h₄ ▸ dvd_add h₁ (dvd_refl _) }
+end
+
+private lemma pow_gt_base (a b : ℕ) (ha : 1 < a) (hb : 1 < b) : a < a^b :=
+begin
+  have ha₁ : 0 < a := pos_of_gt ha,
+  have hb₁ : 1 ≤ b := le_of_lt hb,
+
+  rw [←nat.sub_add_cancel hb₁, pow_succ a (b - 1)],
+  nth_rewrite_lhs 0 ←mul_one a,
+  suffices h : 1 < (a^(b - 1)),
+  { exact mul_lt_mul_of_pos_left h ha₁ },
+  have : 0 < (b - 1) := tsub_pos_of_lt hb,
+  exact (b - 1).one_lt_pow a this ha
+end
```

Comment:
```suggestion
private lemma pow_gt_base (a b : ℕ) (ha : 1 < a) (hb : 1 < b) : a < a^b :=
by convert pow_lt_pow ha hb; simp
```
We essentially have this one already with `pow_lt_pow`.

Reply:
I only used `pow_gt_base` once, so I just removed the lemma altogether and replaced it directly with `pow_lt_pow`.